### ISSUE TITLE
`unnecessary-else` emit warning on `else if`

### ISSUE THREAD

**Issue by TokugawaTakesi** [palantir/tslint#4658]
### Bug Report

- __TSLint version__: 5.15.0
- __TypeScript version__: 3.4.1
- __Running TSLint via__: (pick one) WebStorm + TSLoader(Webpack)

#### TypeScript code being linted

```ts
if (isDevelopmentMode) {
  return outputDirectoryAbsolutePathForDevelopmentBuild;
} else if (isProductionMode) {
  if (!someCondition) {
    throw new Error();
  }
  return outputPath;
}
```

with `tslint.yaml` configuration:

```yaml
rules:

  encoding: true

  quotemark:
    - true
    - single
    - avoid-template

  semicolon:
    - true
    - always

  no-var-keyword: true

  trailing-comma:
  - true
  - multiline: never
    singleline: never

  prefer-const: true

  no-magic-numbers: true

  no-string-literal: true

  object-literal-shorthand:
    - true
    - always

  only-arrow-functions:
    - true
    - allow-named-functions

  no-duplicate-super: true

  no-duplicate-switch-case: true

  deprecation: true
  no-duplicate-imports: true
  no-require-imports: true

  no-unnecessary-initializer: true
  unnecessary-else: true
  no-empty: true

  no-tautology-expression: true

  no-unused-expression:
    - true
    - allow-fast-null-checks
    - allow-new
    - allow-tagged-template

  triple-equals: true
  no-boolean-literal-compare: true
  prefer-object-spread: true
  arrow-return-shorthand: true
  unnecessary-bind: true

  static-this: true
  unnecessary-constructor: true
  no-invalid-this: true
  class-name: true

  no-any: true

  typedef:
    - true
    - call-signature
    - arrow-call-signature
    - property-declaration
    - variable-declaration
    - member-variable-declaration
    - object-destructuring
    - array-destructuring

  ban-types:
    - true
    -
      - 'Number'
    -
      - 'String'
    -
      - 'Boolean'
    -
      - 'Object'

  member-access:
    - true
    - check-accessor

  prefer-readonly:
    - true

  array-type:
    - true
    - generic

  interface-name:
    - true/WebpackConfigHelper
    - always-prefix

  no-redundant-jsdoc: true
```

#### Actual behavior

In `} else if (isProductionMode) {` row, TSLint tells 

> The preceding `if` block ends with a `return` statement. This `else` is unnecessary.

#### Expected behavior

TSLint ignore this row because there is `else if`, not `else`.

**Comment by postama** [comment]
It appears that the `else` is unnecessary. Removing it would create code with the same functionality - since the preceding `if` block ends with `return`.

There is no formal `else if` in Javascript. It is just an `else` statement  followed by an `if` statement.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/if...else#Description

**Comment by TokugawaTakesi** [comment]
Thank you for the answer!

Regardless to JavaScript implementation, 

``` typescript
if (isDevelopmentMode) {
  // ...
} else if (isProductionMode) {
  // ...
}
```

and 

``` typescript
if (isDevelopmentMode) {
  // ...
} else (isProductionMode) {
  // ...
}
```

are different from the point of view of the algorithm. I could be something like

```
if (isDevelopmentMode) {
  // ...
} else if (isProductionMode) {
  // ...
} else if (isTestMode) {
 // 
} 

throw new Error(`Undefined mode`);
```

in first case.

**Comment by postama** [comment]
@TokugawaTakesi 

This rule only runs if the preceding `if` block ends with a `return` statement.
And In that case:
```
if (isDevelopmentMode) {
  // ...
  return 'dev'
} else if (isProductionMode) {
  // ...
  return 'prod'
}
```
is identical to 
```
if(isDevelopmentMode) {
  // ...
  return 'dev'
}
if(isProductionMode) {
  // ...
  return 'prod'
}
```

You are definitely correct that they are different _if the preceding if block does not end with a return_ but since this rule only applies in that case, I think this is a valid rule.

If you can create a test case where the `The preceding if block ends with a return statement. This else is unnecessary.` lint rule shows - and there is a functional difference between using the `if` and `else if` code paths, I would agree it is a bug, and please post it here.

**Comment by JoshuaKGoldberg** [comment]
@postama is correct. #4600 covers the case of what would happen if they were different.
